### PR TITLE
Futures in Scala 2.12: Part 8 — sun.misc.Unsafe

### PR DESCRIPTION
**Pull request by viktorklang**

### CODE REVIEW

**Comment by slandelle** [comment]
Hi Viktor,

This post is very frustrating: you only tell the beginning and the end of the story! What about the journey? :)

`AtomicReference` was there from the beginning, so why originally go with ARFU? How is it that you were capable of getting the same performance with AR? Did AR performance improved in JDK8? 

Cheers, and thanks a lot!

**Comment by viktorklang** [comment]
Hi @slandelle,
Good point, I'll issue an update and try to flesh that out :)
It's not going to be what you think :)

**Comment by slandelle** [comment]
> It's not going to be what you think :)

Oh man, the cliffhanger!

**Comment by viktorklang** [comment]
@slandelle Alright, updated the post to contain more history :)

**Comment by He-Pin** [comment]
@slandelle one thing I know is http://normanmaurer.me/blog/2013/10/28/Lesser-known-concurrent-classes-Part-1/

**Comment by slandelle** [comment]
@hepin1989 Yes, I was aware of Norman's post. So having `AtomicReference` perform better than ARFU would mean raw speed would compensate for additional GC?

**Comment by slandelle** [comment]
> with extending AtomicReference internally

Smart! Thanks @viktorklang

**Comment by viktorklang** [comment]
@slandelle :-) Thank you for helping me improve the blog post, I hope you enjoyed the reading

**Comment by slandelle** [comment]
@viktorklang Absolutely, thanks a lot for sharing! FYI, I just ping'ed the Netty guys: https://github.com/netty/netty/issues/4922. Hunting FieldUpdaters down now :)

**Comment by He-Pin** [comment]
@slandelle http://shipilev.net/blog/2015/faster-atomic-fu/ here is another one.
the real story should waiting for the update,but I think ,Future have only 1 value,that's ok.

**Comment by slandelle** [comment]
@hepin1989 Thanks for the link

**Comment by metasim** [comment]
Why didn't the original implementation take the approach of extending `AtomicReference`? Was it just an evolution in the design, or was it something new made available in Java 8?

**Comment by viktorklang** [comment]
@metasim Hmmm, good question. Perhaps @axel22 knows? https://github.com/scala/scala/blob/ab84c8d9a97b41728e77f7808eda2748d052ca06/src/library/scala/concurrent/impl/AbstractPromise.java

**Comment by axel22** [comment]
Not sure, possibly because performance differences are not so large in real programs. One would have to benchmark large future-heavy applications to measure the difference.

Decisions could date back even earlier to this:

https://github.com/scala/scala/commit/2c01191b99d912f99a7d50be693d09406761a94c#diff-a7acaf0de3e4f9297f5d9b507be3d2cbR64

**Comment by viktorklang** [comment]
Ah! Also, there was the retracted @static annotation that would have
obviated the need for the AbstractPromise class.

## 

Cheers,
√

File: Futures-in-Scala-2.12-part-8.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+A `Future` can be seen as a tri-state<sup>[1](#tristate)</sup> state machine with the following distinct states:
+
+1. Uncompleted
+2. Completed with a successful result
+3. Completed with a failed result
+
+In order to have an «asynchronous Future» it is vital to be able to register logic to be executed once the `Future` becomes Completed.
+
+This means that we can encode the state `1` as a «sequence of callbacks & their `ExeutionContext`», state `2` as a `scala.util.Success`, and state `3` as a `scala.util.Failure`.
```
Comment (ericluria) [inline Futures-in-Scala-2.12-part-8.md:16]: typo here: `ExeutionContext`

great series of blog posts, btw!

Comment (viktorklang) [reply Futures-in-Scala-2.12-part-8.md:16]: @ericluria Thank you so much for both the help and the kind words!

**Pull request merged**